sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java

```diff
@@ -0,0 +1,108 @@
+import org.apache.beam.sdk.coders.RowCoder;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.schemas.SchemaTranslation;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.util.CoderUtils;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.PInput;
+import org.apache.beam.sdk.values.POutput;
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.vendor.grpc.v1p43p2.com.google.common.collect.Iterables;
+import org.apache.beam.vendor.grpc.v1p43p2.com.google.protobuf.ByteString;
+
+/** Wrapper for invoking external Python transforms. */
+public class ExternalPythonTransform<InputT extends PInput, OutputT extends POutput>
+    extends PTransform<InputT, OutputT> {
+  private final String fullyQualifiedName;
+  private final Row args;
+  private final Row kwargs;
+
+  public ExternalPythonTransform(String fullyQualifiedName, Row args, Row kwargs) {
+    this.fullyQualifiedName = fullyQualifiedName;
+    this.args = args;
+    this.kwargs = kwargs;
+  }
+
+  @Override
+  public OutputT expand(InputT input) {
+    int port;
+    try {
+      port = PythonService.findAvailablePort();
+      PythonService service =
+          new PythonService(
+              "apache_beam.runners.portability.expansion_service_main",
+              "--port",
+              "" + port,
+              "--fully_qualified_name_glob",
+              "*");
+      Schema payloadSchema =
+          Schema.of(
+              Schema.Field.of("constructor", Schema.FieldType.STRING),
+              Schema.Field.of("args", Schema.FieldType.row(args.getSchema())),
+              Schema.Field.of("kwargs", Schema.FieldType.row(kwargs.getSchema())));
+      payloadSchema.setUUID(UUID.randomUUID());
+      Row payloadRow =
+          Row.withSchema(payloadSchema).addValues(fullyQualifiedName, args, kwargs).build();
+      ExternalTransforms.ExternalConfigurationPayload payload =
+          ExternalTransforms.ExternalConfigurationPayload.newBuilder()
+              .setSchema(SchemaTranslation.schemaToProto(payloadSchema, true))
+              .setPayload(
+                  ByteString.copyFrom(
+                      CoderUtils.encodeToByteArray(RowCoder.of(payloadSchema), payloadRow)))
+              .build();
+      try (AutoCloseable p = service.start()) {
+        PythonService.waitForPort("localhost", port, 5000);
+        PTransform<PInput, PCollectionTuple> transform =
+            External.<PInput, Object>of(
+                    "beam:transforms:python:fully_qualified_named",
+                    payload.toByteArray(),
+                    "localhost:" + port)
+                .withMultiOutputs();
+        PCollectionTuple outputs;
+        if (input instanceof PCollection) {
+          outputs = (PCollectionTuple) ((PCollection) input).apply(transform);
```

Comment:
Do we need this explicit casting to `(PCollectionTuple)` ?

Reply:
Yes, unfortunately. It seems java struggles with "paritally-raw" types.

Reply:
OK, I was able to work around this by using `PCollection<?>`.